### PR TITLE
Avoid unnecessary indirection in char.ToString

### PR DESCRIPTION
**Pull request by jamesqo**
It's not really necessary to call the string constructor in `char.ToString`; we can just call `FastAllocateString` directly with 1, set `m_firstChar` to the char, and then return it. This makes the method on average more than twice as fast (~0.68 to ~0.31), and it's still being inlined. It also leaves us free to optimize the `string(char, int)` constructor for larger strings w/o worrying about degrading perf for this method, which I may be doing in the future.

cc @jkotas

### CODE REVIEW

File: src/mscorlib/src/System/String.cs [hunk 1]
```diff
@@ -49,7 +49,7 @@ public sealed class String : IComparable, ICloneable, IConvertible, IEnumerable
 
         // For empty strings, this will be '\0' since
         // strings are both null-terminated and length prefixed
-        [NonSerialized] private char m_firstChar;
+        [NonSerialized] internal char m_firstChar;
```
Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: It is pretty unfortunate to make m_firstChar internal for this...

File: src/mscorlib/src/System/Char.cs [hunk 2]
```diff
@@ -163,9 +163,13 @@ public bool Equals(Char obj)
...
       // Provides a string representation of a character.
       [Pure]
-      public static String ToString(char c) {
+      public static String ToString(char c)
+      {
           Contract.Ensures(Contract.Result<String>() != null);
-          return new String(c, 1);
+
+          string result = string.FastAllocateString(1);
+          result.m_firstChar = c;
```
Comment (benaadams) [inline src/mscorlib/src/System/Char.cs]: Use `fixed` to access the first char instead?

Comment (benaadams) [inline src/mscorlib/src/System/Char.cs]: e.g.
Use `fixed` to access the first char instead?
```csharp
string result = string.FastAllocateString(1);
fixed (char* p = result)
{
    p[0] = c;
}
return result;
```

Comment (jamesqo) [inline src/mscorlib/src/System/Char.cs]: @benaadams using `fixed (ptr = str)` generates less efficient code than `fixed (ptr = &str.m_firstChar)`. Also fixed itself also adds a bit of overhead with pinning the object.

I think what I'll do is simply add a method on `String` that does the work, e.g. `String.CreateFromChar(char)`, and defer to that.

File: src/mscorlib/src/System/String.cs (hunk 1 above)
Comment (jamesqo) [inline src/mscorlib/src/System/String.cs]: @jkotas I have changed the access modifier back to private and moved the actual work in the function to String.

File: src/mscorlib/src/System/String.cs [hunk 3]
```diff
@@ -3778,6 +3778,16 @@ internal unsafe static void InternalCopy(String src, IntPtr dest,int len)
                 Buffer.Memcpy(dstPtr, srcPtr, len);
             }
         }      
+
+        // This is only intended to be used by char.ToString.
+        // It is necessary to put the code in this class instead of Char, since m_firstChar is a private member.
+        // Making m_firstChar internal would be dangerous since it would make it much easier to break String's immutability.
+        internal static string CreateFromChar(char c)
```
Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: Nit - it would be nice to put it next to other `static string Create*` methods.

Comment (jamesqo) [inline src/mscorlib/src/System/String.cs]: I put it near the constructors, which are near the `CreateStringFromEncoding` / `CreateString` methods.

Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: Why not in right next to the CreateStringFromEncoding or CreateString methods?

Also, could you please add `[System.Security.SecuritySafeCritical]` security annotations on this method? (It is needed for full framework.)

**Comment by jkotas** [comment]
Thanks!

File: src/mscorlib/src/System/String.cs (hunk 1 above)
Comment (jamesqo) [inline src/mscorlib/src/System/String.cs]: @jkotas, now that `EditorBrowsableAttribute` is in corelib maybe we could mark this with `EditorBrowsableState.Never` and use it directly from places such as `char` / `StringBuilder`?

Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: EditorBrowsable is about how public APIs show in intellisense. It is not interesting for enforcing implementation hygiene that the `private` modifier is used for here.

**Pull request merged**